### PR TITLE
Feature/send invoice reminder multiple times

### PR DESCRIPTION
**Pull request by AmishaBisht**
Targets #1783 

### Description
I have added the feature of sending multiple reminders and storing those reminders in an invoice_mails table.

### Checklist:
I have done the self-review.

### CODE REVIEW

File: Modules/Invoice/Database/Migrations/2022_07_12_060306_drop_reminder_mail_count_from_invoices_table.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('invoices', function (Blueprint $table) {
+            // $table->dropColumn("reminder_mail_count");
```
Comment (GauravGusain98) [inline Modules/Invoice/Database/Migrations/2022_07_12_060306_drop_reminder_mail_count_from_invoices_table.php:17]: I think we need this line

File: Modules/Invoice/Services/InvoiceService.php [hunk 2]
```diff
@@ -560,6 +561,13 @@ public function sendInvoiceReminder(Invoice $invoice, $data)
         $invoice->update([
             'reminder_mail_count' => ($invoice->reminder_mail_count + 1)
         ]);
+        // dd($invoice);
```
Comment (GauravGusain98) [inline Modules/Invoice/Services/InvoiceService.php:564]: remove this

File: Modules/Invoice/Entities/Invoice.php [hunk 3]
```diff
@@ -184,9 +184,13 @@ public function getInvoiceAmountInInrAttribute()
...
+    public function invoiceMail()
+    {
+        return $this->hasMany(InvoiceMail::class, 'invoice_id', 'id');
+    }
 
-    public function getTotalAmountAttribute()
+    public function latestReminder()
     {
-        return $this->amount + $this->gst;
+        return $this->invoiceMail()->orderByDesc('sent_on')->first();
```
Comment (GauravGusain98) [inline Modules/Invoice/Entities/Invoice.php:194]:
```suggestion
        return $this->invoiceMail()->where('type', config('invoice.mail-type.invoice-reminder.slug'))->orderByDesc('sent_on')->first();
```

Comment (GauravGusain98) [reply Modules/Invoice/Entities/Invoice.php:194]: We only want reminder type emails

File: Modules/Invoice/Resources/views/index.blade.php [hunk 4]
```diff
@@ -122,10 +122,8 @@
                                     {{ $invoice->shouldHighlighted() ? __('Overdue') : $invoice->status }}
                                 </td>
                                 <td class="text-center">
-                                    @if($invoice->reminder_mail_count)
-                                        <div class="text-success">{{ __('Reminder Sent') }}</div>
-                                    @elseif($invoice->shouldHighlighted())
-                                        <div class="btn btn-sm btn-primary send-reminder" data-invoice-data="{{ json_encode($invoiceData) }}" data-toggle="modal" data-target="#invoiceReminder" >{{ __('Reminder') }}</div>
+                                    @if($invoice->shouldHighlighted())
+                                        <div class="btn btn-sm btn-primary send-reminder" data-invoice-data="{{ json_encode($invoiceData) }}" data-toggle="modal" data_toogle="tooltip" data-placement="top" title="{{optional($invoice->latestReminder())->sent_on }}" data-target="#invoiceReminder" >{{ __('Reminder') }} ({{$invoice->invoiceMail->count()}})</div>
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/index.blade.php:126]: data-toggle is used 2 times here. Please find a solution to fix this as we need to toggle modal and tooltip from same button

**Review by GauravGusain98 — CHANGES_REQUESTED** [review]
Please check the comments.

File: Modules/Invoice/Services/InvoiceService.php (hunk 2 above)
Comment (amisha296) [reply Modules/Invoice/Services/InvoiceService.php:564]: okay

File: Modules/Invoice/Database/Migrations/2022_07_12_060306_drop_reminder_mail_count_from_invoices_table.php (hunk 1 above)
Comment (amisha296) [reply Modules/Invoice/Database/Migrations/2022_07_12_060306_drop_reminder_mail_count_from_invoices_table.php:17]: okay

File: Modules/Invoice/Resources/views/edit.blade.php [hunk 5]
```diff
@@ -1,15 +1,15 @@
 @extends('invoice::layouts.master')
 @section('content')
 @php
-$filename=basename($invoice->file_path);
+ $filename=basename($invoice->file_path);
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/edit.blade.php:4]:
```suggestion
    $filename=basename($invoice->file_path);
```

File: Modules/Invoice/Resources/views/edit.blade.php [hunk 6]
```diff
@@ -1,15 +1,15 @@
...
 @endphp
 <div class="container">
     <br>
-    
     <div class="d-flex justify-content-between mb-2">
     @if(auth()->user()->can('finance_invoices.update'))
             <h4>Invoice Information</h4>
             <span>
+                <a class="btn btn-sm btn-info text-white mr-4 font-weight-bold" data-toggle="modal" data-target="#previewMails" href="#">Preview mails</a>
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/edit.blade.php:12]: Please use this when you are not using href
**https://www.tutorialspoint.com/How-to-make-an-anchor-tag-refer-to-nothing#:~:text=To%20make%20an%20anchor%20tag%20refer%20to%20nothing%2C%20use%20%E2%80%9Cjavascript,back%20into%20the%20current%20document.**

File: Modules/Invoice/Resources/views/modals/email-details.blade.php [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+                    <span aria-hidden="true">&times;</span>
+                </button>
+            </div>
+            <div class="modal-body">
+                @foreach ($emails as $email)
+                    <div class="card mt-4">
+                        <form action="" method="POST">
+                            {{ csrf_field() }}
+                            {{ method_field('PATCH') }}
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/modals/email-details.blade.php:17]:
```suggestion
                            @csrf
```

File: Modules/Invoice/Resources/views/modals/email-details.blade.php [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                    <h4 class="modal-title">{{ __('Preview Mails') }}</h4>
+                </div>
+                <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+                    <span aria-hidden="true">&times;</span>
+                </button>
+            </div>
+            <div class="modal-body">
+                @foreach ($emails as $email)
+                    <div class="card mt-4">
+                        <form action="" method="POST">
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/modals/email-details.blade.php:15]: remove this  form

File: Modules/Invoice/Resources/views/modals/email-details.blade.php [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+                                                    <textarea type="text" placeholder="{{ $email['body'] }}" rows="10" class=" form-control" disabled></textarea>
+                                                </div>
+                                            </div>
+                                        </div>
+                                    </div>
+                                    <div class="card-footer">
+                                        <button type="submit" class="btn btn-primary">Save</button>
+                                    </div>
+                                </div>
+                        </form>
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/modals/email-details.blade.php:55]: remove this

File: Modules/Invoice/Resources/views/subviews/edit/invoice-details.blade.php [hunk 10]
```diff
@@ -161,7 +161,7 @@
                 </div>
             </div>
             <div v-if="status == 'paid'">
-                @if($invoice->payment_confirmation_mail_sent === 0)
+                @if($invoice->getInvoiceConfirmationMail() === null)
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/subviews/edit/invoice-details.blade.php:164]:
```suggestion
                @if(empty($invoice->getInvoiceConfirmationMail()))
```

File: Modules/Invoice/Services/InvoiceService.php [hunk 11]
```diff
@@ -149,14 +150,17 @@ public function update($data, $invoice)
...
         }
         if (isset($data['invoice_file']) and $data['invoice_file']) {
             $this->saveInvoiceFile($invoice, $data['invoice_file']);
             $this->setInvoiceNumber($invoice, $data['sent_on']);
         }
+        InvoiceMail::create([
+            'invoice_id' => $invoice->id,
+            'body' => $data['email_body'],
+            'subject' => $data['email_subject'],
+            'type' => config('invoice.mail-type.payment-confirmation.label')
```
Comment (GauravGusain98) [inline Modules/Invoice/Services/InvoiceService.php:162]: Please add sent_on time as well. Set it to current time

File: Modules/Invoice/Services/InvoiceService.php [hunk 12]
```diff
@@ -539,6 +544,12 @@ public function sendInvoice(array $data)
         $invoiceNumber = str_replace('-', '', optional($client)->next_invoice_number ?: $project->next_invoice_number);
         $invoice = $this->createInvoice($client, $project, $term);
         Mail::queue(new SendInvoiceMail($invoice, $invoiceNumber, $email));
+        InvoiceMail::create([
+            'invoice_id'=> $invoice->id,
+            'subject' => $email['subject'],
+            'body' => $email['body'],
+            'type' => config('invoice.mail-type.invoice.label')
```
Comment (GauravGusain98) [inline Modules/Invoice/Services/InvoiceService.php:551]: Same here

File: Modules/Invoice/Services/InvoiceService.php [hunk 13]
```diff
@@ -580,6 +591,12 @@ public function sendInvoiceReminder(Invoice $invoice, $data)
         $invoice->update([
             'reminder_mail_count' => ($invoice->reminder_mail_count + 1)
         ]);
+        InvoiceMail::create([
+            'invoice_id'=> $invoice->id,
+            'subject' => $email['subject'],
```
Comment (GauravGusain98) [inline Modules/Invoice/Services/InvoiceService.php:596]: Same here

File: Modules/Invoice/Services/InvoiceService.php [hunk 14]
```diff
@@ -665,7 +682,7 @@ public function createInvoice($client, $project, $term)
         $pdf->generateFromHtml($html, storage_path('app' . $filePath), [], true);
         $invoice->update([
             'invoice_number' => $invoiceNumber,
-            'file_path' => $filePath
+            'file_path' => ''
```
Comment (GauravGusain98) [inline Modules/Invoice/Services/InvoiceService.php:685]: Please fix this as we have removed this just for testing

**Review by GauravGusain98 — CHANGES_REQUESTED** [review]
Please check the comments

**Comment by GauravGusain98** [comment]
Found some issues while testing
1. When sending reminder, getting this error
image

2. After sending the invoice the reminder count shows 1 instead of 0
image

3. In the preview mails, Don't know why there is a save button as we don't need to update anything.
image

**Comment by amisha296** [comment]
@GauravGusain98  I have resolved the issues, kindly check the PR.

**Comment by GauravGusain98** [comment]
@amisha296 The first 2 issues are not yet fixed. Still getting the errors. Found one more issue related to alignment of the sent on time. It should be right aligned instead of starting in between

image

**Comment by GauravGusain98** [comment]
Looks fine will approve and merge it after testing

**Pull request closed without merging**